File: leetcode/2. problem/java/77.组合.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=77 lang=java
 *
 * [77] 组合
 */

// @lc code=start
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        return extracted(n, k);
    }



    
    private List<List<Integer>> extracted(int n, int k) {
        List<List<Integer>> ans = new ArrayList<>();
        backTracking(n, k, ans, new ArrayList<Integer>(), 1);
        return ans;
    }

    private void backTracking(int n,
            int k,
            List<List<Integer>> ans,
            ArrayList<Integer> arr,
            int key) {
        if (arr.size() == k) {
            ans.add(new ArrayList<>(arr));
        } else {
            for (int i = key; i <= n; i++) {
                arr.add(i);
                backTracking(n, k, ans, arr, i + 1);
                arr.remove(arr.size() - 1);
            }
        }
    }
}
// @lc code=end

File: leetcode/2. problem/77.组合.go
/*
 * @lc app=leetcode.cn id=77 lang=golang
 *
 * [77] 组合
 */
package main

// @lc code=start
func combine(n int, k int) [][]int {
	var ans [][]int
	var nums []int
	for i := 1; i <= n; i++ {
		nums = append(nums, i)
	}
	visited := make([]bool, n)
	var comb []int
	var dfs func(idx int)
	dfs = func(idx int) {
		if len(comb) == k {
			ans = append(ans, append([]int(nil), comb...))
		} else {
			for i := idx; i < len(nums); i++ {
				if !visited[i] {
					visited[i] = true
					comb = append(comb, nums[i])
					dfs(i + 1)
					comb = comb[:len(comb)-1]
					visited[i] = false
				}

			}
		}
	}
	dfs(0)
	return ans

}

// @lc code=end
